Add cancel_status param to the cancel order action

aftermath2 opened the pull request:
## What does this PR do?
Fixes #1311

This PR adds the `cancel_status` parameter to the `POST /api/order` endpoint for users to specify the status in which an order should be for the cancellation to take place.

## Checklist before merging
- [x] Install [pre-commit](https://pre-commit.com/#installation) and initialize it: `pip install pre-commit`, then `pre-commit install`. Pre-commit installs git hooks that automatically check the codebase. If pre-commit fails when you commit your changes, please fix the problems it points out.

## Tests
<details><summary>test_cancel_order_cancel_status</summary>

```console
aftermath:~ $ docker exec test-coordinator coverage run manage.py test tests.test_trade_pipeline.TradeTest.test_cancel_order_cancel_status
Creating test database for alias 'default'...
Found 1 test(s).
System check identified no issues (0 silenced).
Regtest network was already ready. Skipping initalization.
.
----------------------------------------------------------------------
Ran 1 test in 5.832s

OK
Destroying test database for alias 'default'...
```

</details>

<details><summary>test_cancel_order_different_cancel_status</summary>

```console
aftermath:~ $ docker exec test-coordinator coverage run manage.py test tests.test_trade_pipeline.TradeTest.test_cancel_order_different_cancel_status
Creating test database for alias 'default'...
Found 1 test(s).
System check identified no issues (0 silenced).
.
----------------------------------------------------------------------
Ran 1 test in 5.898s

OK
Destroying test database for alias 'default'...
Regtest network was already ready. Skipping initalization.
```
</details>

> Some of the tests that weren't modified and are not affected by the changes failed, but I think they should be fixed in a different PR.

aftermath2 commented on api/logics.py:
```diff
@@ -991,7 +991,7 @@ def is_penalized(user):
 
     @classmethod
-    def cancel_order(cls, order, user, state=None):
+    def cancel_order(cls, order, user, cancel_status=None):
```
I used `cancel_status` because `current_status` was a bit confusing. Let me know if I should rename it.

aftermath2 commented on frontend/src/components/TradeBox/index.tsx:
```diff
@@ -244,9 +247,9 @@ const TradeBox = ({ baseUrl, onStartAgain }: TradeBoxProps): JSX.Element => {
   };
 
-  const cancel = function (): void {
+  const cancel = function (no_confirmation: boolean) {
```
If the cancellation needs no confirmation is because it hasn't been taken yet, so we should use `cancel_status` to send the order status we have with the request.

Reckless-Satoshi commented on docs/assets/schemas/api-latest.yaml:
```diff
@@ -469,6 +469,10 @@ paths:
             - `17` - Maker lost dispute
             - `18` - Taker lost dispute
 
+            The client can use `cancel_status` to cancel the order only
```
This doc is unfortunately auto-generated when running test_trade_pipeline, so these strings here will likely be lost I believe. Maybe I am wrong. I think we should add these helpful comments on `oas_schemas.py` just as #1325 and then the autogenerated `api-latest.yaml` will always have this text.

jerryfletcher21 replied:
Yes it is auto generated in `tests/test_api.py` at the start of the tests. Also without `ENUM_NAME_OVERRIDES` `api-latest.yaml` gets really polluted :)

Reckless-Satoshi commented on api/logics.py:
```diff
@@ -1009,10 +1009,17 @@ def cancel_order(cls, order, user, state=None):
             return False, {"bad_request": "You cannot cancel this order"}
 
-        # 1) When maker cancels before bond
+        # 1) If the order status is not the one specified by the user, do not cancel the order.
```
Best to leave this "do not cancel` comments without enumeration. That is, remove the "1)" , and leave all other cases with the number they had. This is consistent with the first cancel rejection case just above this one:
```
        if order.status in do_not_cancel:
            return False, {"bad_request": "You cannot cancel this order"}
```
The intention of enumerating was to have a clear map of the different cancellation conditions and bond slashing because they are all unique in some way.

Reckless-Satoshi commented on api/logics.py:
```diff
@@ -1009,10 +1009,17 @@ def cancel_order(cls, order, user, state=None):
         # The order never shows up on the book and order
         # status becomes "cancelled"
-        if order.status == Order.Status.WFB and order.maker == user:
+        elif order.status == Order.Status.WFB and order.maker == user:
```
let's leave this as it was for simplicity.

Reckless-Satoshi commented on api/logics.py:
```diff
@@ -1009,10 +1009,17 @@ def cancel_order(cls, order, user, state=None):
+        if cancel_status is not None and order.status != cancel_status:
+            return False, {
+                "bad_request": f"Order status {order.status} is not {cancel_status}. " +
+                "The order may have been taken before it was cancelled"
```
This message `"The order may have been taken before it was cancelled"` only applies to a very specific case of `order.status != cancel_status` . Maybe best remove it. Very often this bad_request will be returned because of a check between order status that is not `public != taken`.

jerryfletcher21 replied:
Yes agree, also I think it is better to put this check before the `if order.status in do_not_cancel` like in my pull request, because otherwise if the client has a wrong `cancel_status`, it gets returned "You cannot cancel this trade", but this is referring to the correct order status. I may confuse a bit the client, it is better to first check if the `cancel_status` is correct.

Reckless-Satoshi commented on tests/utils/trade.py:
```diff
@@ -112,11 +112,13 @@ def get_order(self, robot_index=1, first_encounter=False):
         headers = self.get_robot_auth(robot_index)
         body = {"action": "cancel"}
+        if cancel_status is not None:
+            body.update({"cancel_status": cancel_status})
```
A little bit cumbersome. We can leave this simple as `body = {"action": "cancel", "cancel_status": cancel_status}` without need for `if` statement right ?

jerryfletcher21 replied:
It is a small thing, but in this way the test better fits with how the api is actually called by clients in mainnet. The client either specifies a non `None` `cancel_status`, or it does not specifiy `cancel_status` all together, it is quite rare that it will specify `cancel_status` as `None`. It is really a minor thing, can also be changed with the way you specified.

aftermath2 replied:
I used the if condition because of what Jerry mentions, but that does not affect these tests, so I will replace it with your suggestion instead

Reckless-Satoshi replied:
Have no preference really, both views are valid now that I read the thinking behind.

Reckless-Satoshi reviewed (changes requested):
Looks good! I left a few comments for some minor changes.

We can get inspiration from #1325 for 1) the fewer/simpler changes to `logics.py`, 2) the addition of comments into `oas_schemas.py` and 3) `ENUM_NAME_OVERRIDES` as this hack leads into a more complete `api-latest` generated schema.

Thank you for implementing the test cases, this is very important.

We will merge this PR in favor of #1325 but both of them are technically good, so what if we split a bit the rewards? 70K for this PR and 40K for #1325.

jerryfletcher21 commented:
@Reckless-Satoshi I think that `api_specs.yaml` was auto generated in the past but it is no longer used anywhere, I can not see any reference to it in the code, and it is also quite outdated (v0.5.3), I think it can be removed?

jerryfletcher21 commented:
> Looks good! I left a few comments for some minor changes.
> 
> We can get inspiration from #1325 for 1) the fewer/simpler changes to `logics.py`, 2) the addition of comments into `oas_schemas.py` and 3) `ENUM_NAME_OVERRIDES` as this hack leads into a more complete `api-latest` generated schema.
> 
> Thank you for implementing the test cases, this is very important.
> 
> We will merge this PR in favor of #1325 but both of them are technically good, so what if we split a bit the rewards? 70K for this PR and 40K for #1325.

Too many #1325 :) But I understand, perfectly fine by me.

aftermath2 commented on robosats/middleware.py:
```diff
@@ -8,7 +8,6 @@
 from django.utils.deprecation import MiddlewareMixin
 from django.http import JsonResponse
 from rest_framework.authtoken.models import Token
-from rest_framework.exceptions import AuthenticationFailed
```
Removed this line to fix the lint check [error](https://github.com/RoboSats/robosats/actions/runs/9549559934/job/26319402405?pr=1326)

```console
robosats/middleware.py:11:39: F401 [*] `rest_framework.exceptions.AuthenticationFailed` imported but unused
```

because the import is not used, but it failed with the same message in the run of the commit that removed it 🤔

Reckless-Satoshi replied:
It's a linting error that made it to `main` on a merge conflict resolution. Should now be fine if this branch is updated.

Reckless-Satoshi commented:
> @Reckless-Satoshi I think that `tests/api_specs.yaml` was auto generated in the past but it is no longer used anywhere, I can not see any reference to it in the code, and it is also quite outdated (v0.5.3), I think it can be removed?

Yes, in fact, I removed it from `main` yesterday as I also realized this fact :)

Reckless-Satoshi commented:
> Too many #1325 :) But I understand, perfectly fine with me.

Thank you, I actually don't know how to handle this situation 😅😂 Kudos is yours for thinking about sending the expected status. 😁

Reckless-Satoshi commented:
I'll need to test the frontend part before we merge. Will likely not be able until sunday.

aftermath2 commented:
Update: rebased the branch to pull the changes from `main` and run the lint check again.

> I'll need to test the frontend part before we merge. Will likely not be able until sunday.

Alright, no problem 👍

aftermath2 commented:
Update: rebased to the updated base branch and fixed a conflict in the `tests/test_trade_pipeline.py` file.

Reckless-Satoshi commented:
I might be doing something wrong here. Currently 17 tests are failing for me

```python
======================================================================
ERROR: test_withdraw_reward_after_unilateral_cancel (tests.test_trade_pipeline.TradeTest.test_withdraw_reward_after_unilateral_cancel)
Tests withdraw rewards as taker after maker cancels order unilaterally
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/robosats/tests/test_trade_pipeline.py", line 1226, in test_withdraw_reward_after_unilateral_cancel
    trade.cancel_order(trade.maker_index)
  File "/usr/local/lib/python3.12/unittest/mock.py", line 1390, in patched
    return func(*newargs, **newkeywargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/robosats/tests/utils/trade.py", line 120, in cancel_order
    self.response = self.client.post(path + params, body, **headers)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/rest_framework/test.py", line 296, in post
    response = super().post(
               ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/rest_framework/test.py", line 209, in post
    data, content_type = self._encode_data(data, format, content_type)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/rest_framework/test.py", line 180, in _encode_data
    ret = renderer.render(data)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/rest_framework/renderers.py", line 916, in render
    return encode_multipart(self.BOUNDARY, data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/test/client.py", line 300, in encode_multipart
    raise TypeError(
TypeError: Cannot encode None for key 'cancel_status' as POST data. Did you mean to pass an empty string or omit the value
```

aftermath2 commented:
@Reckless-Satoshi That was caused by what we discussed [here](https://github.com/RoboSats/robosats/pull/1326#discussion_r1641986850). I may have executed the tests with a cached environment so I didn't notice the error after I changed it (my bad). I've just updated that part of the code so it sets the `cancel_status` only if it's not `None`.

aftermath2 commented:
@Reckless-Satoshi Just rebased, solved some conflicts and ran the tests to validate that everything works as expected. Please let me know if there's anything else I should do from my side to get this merged.

aftermath2 commented:
@Reckless-Satoshi Just rebased, solved some conflicts and ran the tests (results in the [description](https://github.com/RoboSats/robosats/pull/1326#issue-2354419065)) to validate that everything works as expected. Please let me know if there's anything else I should do from my side to get this merged.

MrImmortal09 commented:
@aftermath2 are you still working on this?

aftermath2 commented:
@MrImmortal09 Yes, just waiting for it to be reviewed 👍

KoalaSat commented on api/logics.py:
```diff
@@ -996,7 +996,17 @@ def is_penalized(user):
+        # if it is not the correct one.
+        # This prevents the client from cancelling an order that
+        # recently changed status.
+        if cancel_status is not None:
```
I believe this is going to be difficult to deploy given our decentralized structure, we have multiple versions of clients on multiple kind of devices, and also multiple coordinators who won't update at the same time. This change has be soft.

aftermath2 replied:
Is there any way we can request the server version from the client to decide whether to include `cancel_status` in the request? That way we could avoid issues on coordinators with older versions.

Because to be retro-compatible we would have to use an existing parameter like `statement` to specify the `cancel_status`, but that doesn't seem ideal.

KoalaSat replied:
Why not just have cancel_status as optional and let people slowly migrate? Once de feel safe de can just make it mandatory.

Its either orders with cancel_estatus empty (at the beggning) or a lot users with unusable apps

aftermath2 replied:
Yes, `cancel_status` is already optional on the server side because it's not marked as `required` in the API spec and has a default value of `None`.

But what do we do in the client? Do we skip it altogether until all coordinators have migrated? Because if we specify the parameter, the request will fail on older coordinators.

KoalaSat replied:
I think the best option is to keepnit optional until we have a version of the frontend compatible with the new attribute available in all platforms (mobile,desktop,seflhosrt...) So if an user has issues with an old version they can easily upgrade.

aftermath2 replied:
On the server side it's already optional, I don't think we should change anything.

On the client side it's more tricky. Right now the parameter is only set if the user is the order maker and the cancellation requires no confirmation (the order hasn't been taken yet)

```ts
const noConfirmation = Boolean(order?.is_maker && [0, 1, 2].includes(order?.status));
...
        cancel_status: noConfirmation ? order?.status : undefined
```

By optional do you mean letting the user decide when to send `cancel_status` in these cases? If so, how do we let them decide?

Adding a button just for this parameter is not great UX and they will still get errors if they choose to include it and the coordinator is outdated.

The only clean solution I can think of is adding the parameter in the server in this PR and add it to the clients in a later PR when we are sure most coordinators have them.

KoalaSat replied:
You made me realize it might overcomplicate things. Let's then keep it as it is and I'll think about that for another PR, thanks

aftermath2 replied:
Okay, just rebased to main and resolved some conflicts to leave it ready to review 👍

KoalaSat commented:
@aftermath2 Seems like tests still fails 

```
======================================================================
ERROR: test_cancel_order_cancel_status (tests.test_trade_pipeline.TradeTest.test_cancel_order_cancel_status)
Tests the cancellation of a public order using cancel_status.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/robosats/tests/test_trade_pipeline.py", line 1124, in test_cancel_order_cancel_status
    data["bad_request"], "This order has been cancelled by the maker"
    ~~~~^^^^^^^^^^^^^^^
KeyError: 'bad_request'

----------------------------------------------------------------------
Ran 1 test in 5.783s

```

aftermath2 commented:
@KoalaSat There was a behavior change that caused the `data` variable to not hold the field `bad_request`. I changed it to access the bad request message like other tests do and now both of the ones added run successfully.

```console
aftermath:~ $ docker exec test-coordinator coverage run manage.py test tests.test_trade_pipeline.TradeTest.test_cancel_order_cancel_status
Creating test database for alias 'default'...
Found 1 test(s).
System check identified no issues (0 silenced).
Updating order with new Locked bond from maker
Regtest network was already ready. Skipping initalization.
2025-03-30 00:11:41.415953+00:00
{'num_active_invoices': 1, 'invoices': [{0: {'payment_hash': '92c0f53024e789e7506bc8041d1105395116f06fbbc2f680cd2c30df428d93c8', 'old_status': 'Generated', 'new_status': 'Locked'}}], 'time': 0.04764270782470703}
.
----------------------------------------------------------------------
Ran 1 test in 4.290s

OK
Destroying test database for alias 'default'...
aftermath:~ $ docker exec test-coordinator coverage run manage.py test tests.test_trade_pipeline.TradeTest.test_cancel_order_different_cancel_status
Creating test database for alias 'default'...
Found 1 test(s).
System check identified no issues (0 silenced).
Updating order with new Locked bond from maker
Regtest network was already ready. Skipping initalization.
2025-03-30 00:12:04.577482+00:00
{'num_active_invoices': 1, 'invoices': [{0: {'payment_hash': '1b378e15fbf2e41792c489fb7a0687e0eaa5a839327a62145af31d83d6e1739f', 'old_status': 'Generated', 'new_status': 'Locked'}}], 'time': 0.04600787162780762}
.
----------------------------------------------------------------------
Ran 1 test in 4.962s

OK
Destroying test database for alias 'default'...
```

My bad, I thought they would pass because there were no functional changes since the last run (PR desc), but I should have double checked it.

> There are two other tests failing `test_basic_frontend_url_content` and `test_pro_frontend_url_content` but they are not related to the changes from this PR.

KoalaSat commented:
> @KoalaSat There was a behavior change that caused the `data` variable to not hold the field `bad_request`. I changed it to access the bad request message like other tests do and now both of the ones added run successfully.
> 
> ```
> aftermath:~ $ docker exec test-coordinator coverage run manage.py test tests.test_trade_pipeline.TradeTest.test_cancel_order_cancel_status
> Creating test database for alias 'default'...
> Found 1 test(s).
> System check identified no issues (0 silenced).
> Updating order with new Locked bond from maker
> Regtest network was already ready. Skipping initalization.
> 2025-03-30 00:11:41.415953+00:00
> {'num_active_invoices': 1, 'invoices': [{0: {'payment_hash': '92c0f53024e789e7506bc8041d1105395116f06fbbc2f680cd2c30df428d93c8', 'old_status': 'Generated', 'new_status': 'Locked'}}], 'time': 0.04764270782470703}
> .
> ----------------------------------------------------------------------
> Ran 1 test in 4.290s
> 
> OK
> Destroying test database for alias 'default'...
> aftermath:~ $ docker exec test-coordinator coverage run manage.py test tests.test_trade_pipeline.TradeTest.test_cancel_order_different_cancel_status
> Creating test database for alias 'default'...
> Found 1 test(s).
> System check identified no issues (0 silenced).
> Updating order with new Locked bond from maker
> Regtest network was already ready. Skipping initalization.
> 2025-03-30 00:12:04.577482+00:00
> {'num_active_invoices': 1, 'invoices': [{0: {'payment_hash': '1b378e15fbf2e41792c489fb7a0687e0eaa5a839327a62145af31d83d6e1739f', 'old_status': 'Generated', 'new_status': 'Locked'}}], 'time': 0.04600787162780762}
> .
> ----------------------------------------------------------------------
> Ran 1 test in 4.962s
> 
> OK
> Destroying test database for alias 'default'...
> ```
> 
> My bad, I thought they would pass because there were no functional changes since the last run (PR desc), but I should have double checked it.
> 
> > There are two other tests failing `test_basic_frontend_url_content` and `test_pro_frontend_url_content` but they are not related to the changes from this PR.

thank you :smiley:
